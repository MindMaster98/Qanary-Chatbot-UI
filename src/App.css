#root {
  position: absolute;
  height: 100%;
  width: 100%;
  left: 0%;
  top: 0%;
  display: flex;
  flex-direction: column;
  z-index: 0;
}

#root * {
  font-family: Impact, Haettenschweiler, "Arial Narrow Bold", sans-serif;
}
