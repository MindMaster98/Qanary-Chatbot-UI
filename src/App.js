import { Component } from "react";
import "./App.css";
import { MessageInput, MessagePanel, PageHeader } from "./components";
import chatBotService from "./services/chatbot.service";
import robot_icon_black from "./components/share/imgs/robot_icon_black.svg";

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      messages: [],
      isSending: false,
    };

    this.sendMessage = this.sendMessage.bind(this);
  }

  async sendMessage(messageText = "") {
    this.setState({ isSending: true });
    let messagesCopy = this.state.messages;
    let now = new Date();
    messagesCopy.push({
      text: messageText,
      time: now.getHours() + ":" + ("0" + now.getMinutes()).slice(-2),
      isReply: false,
      loadedSuccessfully: true,
    });
    let reply = await chatBotService.postQuery(messageText);
    if (!!reply.visualization?.buttons)
      reply.visualization.buttons = reply.visualization.buttons.map(
        (button) => ({
          onClick: () => this.sendMessage(button.payload),
          ...button,
        })
      );

    now = new Date();
    messagesCopy.push({
      text: reply.answer,
      followUpNeeded: reply.followUpNeeded,
      loadedSuccessfully: reply.loadedSuccessfully,
      visualization: reply.visualization,
      time: now.getHours() + ":" + ("0" + now.getMinutes()).slice(-2),
      isReply: true,
      icon: robot_icon_black,
    });
    this.setState({
      messages: messagesCopy,
      isSending: false,
    });
  }

  render() {
    return (
      <>
        <PageHeader sendMessage={this.sendMessage} />
        <MessagePanel messages={this.state.messages} />
        <MessageInput
          sendMessage={this.sendMessage}
          isSending={this.state.isSending}
        />
      </>
    );
  }
}

export default App;
