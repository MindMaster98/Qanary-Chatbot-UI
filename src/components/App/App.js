import { Component } from "react";
import "./App.css";
import { MessageInput, MessagePanel, PageHeader } from "..";
import chatBotService from "../../services/chatbot.service";
import robot_icon from "../share/imgs/robot_icon.svg";
import user_icon from "../share/imgs/account_icon_black.svg";
import { textsHelper } from "../../helpers";
import CONFIG from "../../config.json";

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      messages: [],
      components: CONFIG["default-chatbot-components"].map((componentName) => ({
        name: componentName,
        activated: false,
      })),
      backendUrl: "",
      isSending: false,
    };

    this.texts = textsHelper.getTexts();

    this.sendMessage = this.sendMessage.bind(this);
    this.setBackendUrl = this.setBackendUrl.bind(this);
    this.setComponents = this.setComponents.bind(this);
    this.toggleComponent = this.toggleComponent.bind(this);
  }

  componentDidMount() {
    const queryParams = window.location;
    // This RegExp accepts either ?question=questionText or &question=questionText to be as flexible as possible
    const queryRegExp = new RegExp(
      `\\?${CONFIG["initial-question-parameter-name"] || "question"}=([^&]*)|&${
        CONFIG["initial-question-parameter-name"] || "question"
      }=([^&]*)`
    );
    const potentialQueries = queryRegExp.exec(queryParams);
    // Pick the first result by default, unless it's empty or undefined (only possible cases)
    // Then decode it to use the intended string
    const initialQuestionText = decodeURIComponent(
      potentialQueries?.[1] || potentialQueries?.[2] || ""
    );
    if (!!initialQuestionText)
      this.sendMessage(decodeURIComponent(initialQuestionText));
    this.listenerId = textsHelper.addListener(() => {
      this.texts = textsHelper.getTexts();
      this.forceUpdate();
    });
    window.addEventListener("popstate", (popStateEvent) => {
      // Update component state with messages from history state
      if (popStateEvent.state == null) {
        this.setState({ messages: [] });
      } else {
        this.setState({ messages: popStateEvent?.state?.messages ?? [] });
      }
    });
  }

  componentWillUnmount() {
    textsHelper.removeListener(this.listenerId);
  }

  async sendMessage(messageText = "") {
    // push new message to state
    this.setState({ isSending: true });
    let messagesCopy = this.state.messages;
    let now = new Date(Date.now());
    messagesCopy.push({
      text: messageText,
      time: now.getHours() + ":" + ("0" + now.getMinutes()).slice(-2),
      isReply: false,
      loadedSuccessfully: true,
      icon: user_icon,
    });
    // await reply and push it to state
    const preparedComponents = this.state.components
      .filter((component) => component.activated)
      .map((component) => component.name);
    console.log(preparedComponents);
    let reply = await chatBotService.postQuery(
      messageText,
      this.state.backendUrl,
      preparedComponents
    );
    if (!!reply.visualization?.buttons)
      reply.visualization.buttons = reply.visualization.buttons.map(
        (button) => ({
          onClick: () =>
            this.sendMessage(button.payload, this.state.backendUrl),
          ...button,
        })
      );

    now = new Date(Date.now());
    messagesCopy.push({
      text: reply.answer,
      followUpNeeded: reply.followUpNeeded,
      loadedSuccessfully: reply.loadedSuccessfully,
      visualization: reply.visualization,
      time: now.getHours() + ":" + ("0" + now.getMinutes()).slice(-2),
      isReply: true,
      icon: robot_icon,
    });
    this.setState({
      messages: messagesCopy,
      isSending: false,
    });
    // Push new state to history, see https://developer.mozilla.org/en-US/docs/Web/API/History_API
    let url = new URL(window.location);
    url.searchParams.set(
      CONFIG["initial-question-parameter-name"] || "question",
      encodeURIComponent(messageText)
    );
    messagesCopy = messagesCopy.map((message) =>
      message.followUpNeeded
        ? { ...message, visualization: undefined }
        : message
    );
    window.history.pushState(
      {
        path: url.href,
        messages: messagesCopy,
      },
      "",
      url
    );
  }

  setComponents(components) {
    // is array and only contains strings, if it has elements
    if (
      !Array.isArray(components) ||
      (Array.isArray(components) &&
        components.length > 0 &&
        !components.every(
          (component) =>
            component?.name instanceof String ||
            typeof component?.name === "string"
        ))
    ) {
      throw new Error("components have to be an array");
    }

    this.setState({
      components,
    });
  }

  toggleComponent(component) {
    const components = this.state.components;

    const index = components.indexOf(component);

    components[index] = {
      name: component.name,
      activated: !component.activated,
    };

    this.setState({ components });
  }

  async setBackendUrl(backendUrl) {
    // is string and ends with port and optionally a slash and NOT /#/application or something else
    if (
      (typeof backendUrl !== "string" && !(backendUrl instanceof String)) ||
      ((typeof backendUrl === "string" || backendUrl instanceof String) &&
        !/^https{0,1}:\/\/.+?:\d{1,5}\/?$/.test(backendUrl))
    ) {
      throw new Error("backend url needs to be string of url with base route");
    }

    const componentNames =
      (await chatBotService.getComponents(backendUrl)) ?? [];
    const components = componentNames.map((componentName) => ({
      name: componentName,
      activated: false,
    }));

    this.setState({
      backendUrl,
      components,
    });
  }

  render() {
    const now = new Date(Date.now());
    const initialMessage =
      !!this.state.backendUrl && !!this.state.components?.length > 0
        ? this.texts["default-responses"]["initial-message"][
            "is-configured"
          ].replace("{{url}}", this.state.backendUrl)
        : this.texts["default-responses"]["initial-message"][
            "is-not-configured"
          ];
    return (
      <>
        <PageHeader
          sendMessage={this.sendMessage}
          toggleComponent={this.toggleComponent}
          setBackendUrl={this.setBackendUrl}
          setComponents={this.setComponents}
          components={this.state.components}
          backendUrl={this.state.backendUrl}
        />
        <MessagePanel
          messages={[
            {
              text: initialMessage,
              followUpNeeded: false,
              loadedSuccessfully: false, // This is to prevent the source of data bit from showing up
              visualization: {},
              time: now.getHours() + ":" + ("0" + now.getMinutes()).slice(-2),
              isReply: true,
              icon: robot_icon,
            },
            ...this.state.messages,
          ]}
        />
        <MessageInput
          sendMessage={this.sendMessage}
          isSending={this.state.isSending}
        />
      </>
    );
  }
}

export default App;
