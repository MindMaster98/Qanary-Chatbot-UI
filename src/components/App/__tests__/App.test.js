import renderer from "react-test-renderer";
import App from "../App.js";
import chatBotService from "../../../services/chatbot.service.js";

// Snapshot tests
jest.mock("../../../services/chatbot.service.js");

it("renders App correctly", () => {
  // This is important to always have the same date when creating a new snapshot
  const oldNow = Date.now;
  Date.now = () => new Date(0);
  const tree = renderer.create(<App />).toJSON();
  Date.now = oldNow;
  expect(tree).toMatchSnapshot();
});

// Unit tests
// The only thing we can reasonably test is the sending of the message.

it("sends a message", () => {
  // Components are classes/functions, so we can use this to test methods directly
  // React will complain, but we don't care.
  const app = new App();

  app.sendMessage("test");
  expect(chatBotService.postQuery).toHaveBeenCalledWith("test");
  jest.unmock("../../../services/chatbot.service.js");
})