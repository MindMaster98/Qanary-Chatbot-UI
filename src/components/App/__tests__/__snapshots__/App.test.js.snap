// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App correctly 1`] = `
Array [
  <div
    id="header"
    tabIndex="-1"
  >
    <div
      className="clickableIcon "
      onClick={[Function]}
      style={
        Object {
          "maxHeight": "40px",
          "maxWidth": "100px",
          "position": "relative",
          "top": "calc((100% - 40px) / 2)",
        }
      }
      tabIndex="0"
    >
      <img
        alt="Informationen"
        src="info_icon_white.svg"
        tabIndex="-1"
      />
    </div>
    <div
      className="pageTitle"
      tabIndex="0"
    >
      <a
        className="center white-text"
        href="http://coronabot.ins.hs-anhalt.de/"
        rel="noreferrer"
        target="_self"
      >
        Coronabot Demonstrator
      </a>
    </div>
    <div
      className="questionsToggle"
      onClick={[Function]}
      tabIndex="-1"
    >
      <span
        className="center white-text"
      >
        Beispiele
      </span>
    </div>
    <div
      className="clickableIcon "
      onClick={[Function]}
      style={
        Object {
          "maxHeight": "40px",
          "maxWidth": "100px",
          "position": "relative",
          "top": "calc((100% - 40px) / 2)",
        }
      }
      tabIndex="0"
    >
      <img
        alt="Sprache wechseln"
        src="flag_icon.webp"
        tabIndex="-1"
      />
    </div>
  </div>,
  <div
    className="contentContainer hidden"
    dangerouslySetInnerHTML={
      Object {
        "__html": "Diese Anwendung wurde am <a href=\\"https://www.hs-anhalt.de/hochschule-anhalt/fachbereich-5/ueberblick.html\\" target=\\"_blank\\" rel=\\"noreferrer\\">Fachbereich Informatik und Sprachen</a> der <a href=\\"https://hs-anhalt.de/\\" target=\\"_blank\\" rel=\\"noreferrer\\">Hochschule Anhalt</a> im Rahmen der Lehrveranstaltung &quot;Question Answering &amp; Chatbots&quot; unter der Leitung von <a href=\\"https://www.hs-anhalt.de/hochschule-anhalt/service/personenverzeichnis/prof-dr-andreas-both.html\\" target=\\"_blank\\" rel=\\"noreferrer\\">Prof. Dr. Andreas Both</a> entwickelt. Die Betreuung erfolgte durch Prof. Both und <a href=\\"https://www.hs-anhalt.de/hochschule-anhalt/service/personenverzeichnis/aleksandr-perevalov.html\\" target=\\"_blank\\" rel\\"noreferrer\\">Aleksandr Perevalov</a> sowie durch die Projektpartner des <a href=\\"https://www.bmi.bund.de/\\" target=\\"_blank\\" rel=\\"noreferrer\\">Bundesministerium des Innern, für Bau und Heimat</a> und <a href=\\"https://www.itzbund.de/\\" target=\\"_blank\\" rel=\\"noreferrer\\">ITZBund</a>. Es handelt sich hierbei um einen wissenschaftlichen Demonstrator, der offizielle deutsche COVID-19 Chatbot ist <a href=\\"https://chatbot.it.bund.de/\\" target=\\"_blank\\" rel=\\"noreferrer\\">hier</a> abrufbar.<p><h3>Projektteam</h3>Das studentische Projektteam bestand aus Studierenden unserer Studieng&auml;nge &quot;Angewandte Informatik&quot; (Bachelor) und &quot;Data Science&quot; (Master): <ul><li>Backend: Johannes Richard Bartsch, Paul Heinze, Rostislav Iudin, Johannes Rudolf Herkner and Tim Schrader<li>Frontend: Jonas Wunsch</ul>",
      }
    }
    id="imprint"
    onBlur={[Function]}
    tabIndex="0"
  />,
  <div
    className="contentContainer hidden"
    id="exemplaryQuestions"
    onBlur={[Function]}
    tabIndex="0"
  >
    <div
      className="exampleQuestion"
    >
      <div>
        Tote Köln
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Infektionszahlen Frankfurt
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Wie viele Fälle gab es im Februar in Berlin Spandau?
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Fälle 12.02.2021
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Wie viele Fälle wurden diese Woche gemeldet?
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Infektionszahlen Januar
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Todeszahlen Oktober
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Wie viele Tote gab es im Februar in Quedlinburg?
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Fälle gestern
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
    <div
      className="exampleQuestion"
    >
      <div>
        Wie viele Fälle wurden im Juni in Berlin und Brandenburg gemeldet?
      </div>
      <button
        className="button "
        onClick={[Function]}
        style={Object {}}
      >
        Diese Frage stellen
      </button>
    </div>
  </div>,
  <div
    id="messagePanel"
    tabIndex="-1"
  >
    <div
      tabIndex="-1"
    >
      <div
        className="messageContainer reply"
        tabIndex="-1"
      >
        <div
          className="clickableIcon "
          style={
            Object {
              "cursor": "initial",
              "height": "100px",
              "width": "100px",
            }
          }
          tabIndex="-1"
        >
          <img
            alt="Coronabot"
            src="robot_icon.svg"
            tabIndex="-1"
          />
        </div>
        <div
          className="messageDataContainer"
          tabIndex="0"
        >
          <div
            className="messageText white-text"
            dangerouslySetInnerHTML={
              Object {
                "__html": "Hallo, ich bin Coronabot. Ich beantworte Fragen zu den Fallzahlen des Coronavirus bzw. COVID-19 in Deutschland. Ich nutze zwar die offiziellen Datenquellen des <a rel=\\"noreferrer\\" target=\\"_blank\\" href=\\"https://www.rki.de/\\">Robert Koch-Instituts</a>, aber ich bin keine offizielle Anwendung, sondern ein wissenschaftlicher Demonstrator. Der offizielle deutsche Chatbot zum Thema COVID-19 ist unter <a rel=\\"noreferrer\\" target=\\"_blank\\" href=\\"https://chatbot.it.bund.de/\\">chatbot.it.bund.de</a> abrufbar. Dieser Demonstrator wurde an der <a rel=\\"noreferrer\\" target=\\"_blank\\" href=\\"https://hs-anhalt.de/\\">Hochschule Anhalt</a> entwickelt. Mehr Informationen dazu werden im Bereich <img src=\\"imgs/info_icon_white.svg\\" alt=\\"Informationen\\" style=\\"cursor: text\\"></img> dargestellt.",
              }
            }
          />
          <div
            className="messageTimeSent"
          >
            1:00
          </div>
        </div>
      </div>
      <div
        className="richContentContainer"
        tabIndex="-1"
      >
        <div
          className="buttonsContainer"
          tabIndex="-1"
        />
      </div>
    </div>
  </div>,
  <div
    className=""
    id="loadingBar"
    tabIndex="-1"
  />,
  <div
    id="messageInput"
    tabIndex="-1"
  >
    <textarea
      className="black-text"
      id="inputTextBox"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Ihre Frage an mich..."
      value=""
    />
    <div
      id="submitButtonContainer"
      tabIndex="-1"
    >
      <button
        className="button light-gray black-text vertical-center right-side"
        onClick={[Function]}
        style={
          Object {
            "height": "40px",
            "marginRight": "10px",
            "width": "100px",
          }
        }
      >
        Senden
      </button>
    </div>
  </div>,
]
`;
